plugin/cshared.go

```diff
@@ -0,0 +1,284 @@
+package plugin
+
+/*
+#include <stdlib.h>
+*/
+import "C"
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"reflect"
+	"time"
+	"unsafe"
+
+	"github.com/fluent/fluent-bit-go/input"
+	"github.com/fluent/fluent-bit-go/output"
+	"github.com/ugorji/go/codec"
+)
+
+var unregister func()
+
+//export FLBPluginRegister
+func FLBPluginRegister(def unsafe.Pointer) int {
+	defer registerWG.Done()
+
+	if theInput == nil && theOutput == nil {
+		fmt.Fprintf(os.Stderr, "no input or output registered\n")
+		return input.FLB_RETRY
+	}
+
+	if theInput != nil {
+		out := input.FLBPluginRegister(def, theName, theDesc)
+		unregister = func() {
+			input.FLBPluginUnregister(def)
+		}
+		return out
+	}
+
+	out := output.FLBPluginRegister(def, theName, theDesc)
+	unregister = func() {
+		output.FLBPluginUnregister(def)
+	}
+
+	return out
+}
+
+//export FLBPluginInit
+func FLBPluginInit(ptr unsafe.Pointer) int {
+	defer initWG.Done()
+
+	registerWG.Wait()
+
+	if theInput == nil && theOutput == nil {
+		fmt.Fprintf(os.Stderr, "no input or output registered\n")
+		return input.FLB_RETRY
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	var err error
+	if theInput != nil {
+		conf := &flbInputConfigLoader{ptr: ptr}
+		err = theInput.Init(ctx, conf)
+	} else {
+		conf := &flbOutputConfigLoader{ptr: ptr}
+		err = theOutput.Init(ctx, conf)
+	}
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "init: %v\n", err)
+		return input.FLB_ERROR
+	}
+
+	return input.FLB_OK
+}
+
+//export FLBPluginInputCallback
+func FLBPluginInputCallback(data *unsafe.Pointer, csize *C.size_t) int {
+	initWG.Wait()
+
+	if theInput == nil {
+		fmt.Fprintf(os.Stderr, "no input registered\n")
+		return input.FLB_RETRY
+	}
+
+	var err error
+	once.Do(func() {
+		runCtx, runCancel = context.WithCancel(context.Background())
+		theChannel = make(chan Message, 1)
+		go func() {
+			err = theInput.Collect(runCtx, theChannel)
+		}()
+	})
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "run: %s\n", err)
+		return input.FLB_ERROR
+	}
+
+	select {
+	case msg, ok := <-theChannel:
+		if !ok {
+			return input.FLB_OK
+		}
+
+		t := input.FLBTime{Time: msg.Time}
+		b, err := input.NewEncoder().Encode([]any{t, msg.Record})
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "encode: %s\n", err)
+			return input.FLB_ERROR
+		}
+
+		cdata := C.CBytes(b)
+
+		*data = cdata
+		*csize = C.size_t(len(b))
+
+		// C.free(unsafe.Pointer(cdata))
```

Comment:
I think this is been freed by the plugin mechanism, but worth doing a plugin run under valgrind (?)

Reply:
I need help to check those. That's something I don't understand at all.